File: csharp/refactoring-methods/07-generalization-problems.cs
using System;
using System.Collections.Generic;
using System.Linq;

/// <summary>
/// 57. Lifting the field (Pull Up Field)
///
/// BEFORE: Duplicate fields in subclasses
/// </summary>
public class EmployeePullBefore
{
    protected string name;
}

public class ManagerPullBefore : EmployeePullBefore
{
    protected string name; // Duplicate
    private double budget;
}

public class EngineerPullBefore : EmployeePullBefore
{
    protected string name; // Duplicate
    private List<string> skills;
}

/// <summary>
/// AFTER: Pull up field to superclass
/// </summary>
public class EmployeePullAfter
{
    protected string name;
}

public class ManagerPullAfter : EmployeePullAfter
{
    private double budget;
}

public class EngineerPullAfter : EmployeePullAfter
{
    private List<string> skills;
}

/// <summary>
/// 58. Lifting the method (Pull Up Method)
///
/// BEFORE: Duplicate methods in subclasses
/// </summary>
public abstract class ShapeMethodBefore
{
    public abstract double Area();
}

public class CircleMethodBefore : ShapeMethodBefore
{
    private double radius;

    public override double Area()
    {
        return Math.PI * radius * radius;
    }

    public double Circumference()
    {
        return 2 * Math.PI * radius;
    }
}

public class SquareMethodBefore : ShapeMethodBefore
{
    private double side;

    public override double Area()
    {
        return side * side;
    }

    public double Circumference() // Duplicate logic
    {
        return 4 * side;
    }
}

/// <summary>
/// AFTER: Pull up method to superclass
/// </summary>
public abstract class ShapeMethodAfter
{
    public abstract double Area();
    public abstract double Circumference();
}

public class CircleMethodAfter : ShapeMethodAfter
{
    private double radius;

    public override double Area()
    {
        return Math.PI * radius * radius;
    }

    public override double Circumference()
    {
        return 2 * Math.PI * radius;
    }
}

public class SquareMethodAfter : ShapeMethodAfter
{
    private double side;

    public override double Area()
    {
        return side * side;
    }

    public override double Circumference()
    {
        return 4 * side;
    }
}

/// <summary>
/// 59. Lifting the constructor Body (Pull Up Constructor Body)
///
/// BEFORE: Duplicate constructor code
/// </summary>
public class VehicleConstructorBefore
{
    protected string make;
    protected string model;
    protected int year;
}

public class CarConstructorBefore : VehicleConstructorBefore
{
    private int doors;

    public CarConstructorBefore(string make, string model, int year, int doors)
    {
        this.make = make; // Duplicate
        this.model = model; // Duplicate
        this.year = year; // Duplicate
        this.doors = doors;
    }
}

public class TruckConstructorBefore : VehicleConstructorBefore
{
    private double payload;

    public TruckConstructorBefore(string make, string model, int year, double payload)
    {
        this.make = make; // Duplicate
        this.model = model; // Duplicate
        this.year = year; // Duplicate
        this.payload = payload;
    }
}

/// <summary>
/// AFTER: Pull up constructor body
/// </summary>
public class VehicleConstructorAfter
{
    protected string make;
    protected string model;
    protected int year;

    public VehicleConstructorAfter(string make, string model, int year)
    {
        this.make = make;
        this.model = model;
        this.year = year;
    }
}

public class CarConstructorAfter : VehicleConstructorAfter
{
    private int doors;

    public CarConstructorAfter(string make, string model, int year, int doors)
        : base(make, model, year)
    {
        this.doors = doors;
    }
}

public class TruckConstructorAfter : VehicleConstructorAfter
{
    private double payload;

    public TruckConstructorAfter(string make, string model, int year, double payload)
        : base(make, model, year)
    {
        this.payload = payload;
    }
}

/// <summary>
/// 60. Method Descent (Push Down Method)
///
/// BEFORE: Method in wrong class hierarchy level
/// </summary>
public class AnimalPushBefore
{
    public virtual string Speak()
    {
        // Generic implementation
        return "";
    }
}

public class DogPushBefore : AnimalPushBefore
{
    public override string Speak()
    {
        return "Woof";
    }
}

public class CatPushBefore : AnimalPushBefore
{
    public override string Speak()
    {
        return "Meow";
    }
}

public class FishPushBefore : AnimalPushBefore
{
    // Fish don't speak, but inherits speak method
}

/// <summary>
/// AFTER: Push down method to appropriate subclasses
/// </summary>
public class AnimalPushAfter
{
    // No speak method here
}

public class DogPushAfter : AnimalPushAfter
{
    public string Speak()
    {
        return "Woof";
    }
}

public class CatPushAfter : AnimalPushAfter
{
    public string Speak()
    {
        return "Meow";
    }
}

public class FishPushAfter : AnimalPushAfter
{
    // No speak method - appropriate for Fish
}

/// <summary>
/// 61. Field Descent (Push Down Field)
///
/// BEFORE: Field in wrong hierarchy level
/// </summary>
public class EmployeeFieldBefore
{
    protected double salary; // Not all employees have salary
}

public class SalariedEmployeeFieldBefore : EmployeeFieldBefore
{
    // Uses salary
}

public class ContractorFieldBefore : EmployeeFieldBefore
{
    // Doesn't use salary, but inherits it
}

/// <summary>
/// AFTER: Push down field
/// </summary>
public class EmployeeFieldAfter
{
    // No salary field
}

public class SalariedEmployeeFieldAfter : EmployeeFieldAfter
{
    protected double salary;
}

public class ContractorFieldAfter : EmployeeFieldAfter
{
    protected double hourlyRate;
}

/// <summary>
/// 62. Subclass extraction (Extract Subclass)
///
/// BEFORE: Class with conditional behavior
/// </summary>
public class JobExtractBefore
{
    private string type;
    private double rate;
    private double? commission;

    public JobExtractBefore(string type, double rate, double? commission = null)
    {
        this.type = type;
        this.rate = rate;
        this.commission = commission;
    }

    public double GetPay()
    {
        if (type == "salaried")
        {
            return rate;
        }
        else
        {
            return rate + (commission ?? 0);
        }
    }
}

/// <summary>
/// AFTER: Extract subclass
/// </summary>
public abstract class JobExtractAfter
{
    protected double rate;

    public JobExtractAfter(double rate)
    {
        this.rate = rate;
    }

    public abstract double GetPay();
}

public class SalariedJob : JobExtractAfter
{
    public SalariedJob(double rate) : base(rate) { }

    public override double GetPay()
    {
        return rate;
    }
}

public class CommissionedJob : JobExtractAfter
{
    private double commission;

    public CommissionedJob(double rate, double commission) : base(rate)
    {
        this.commission = commission;
    }

    public override double GetPay()
    {
        return rate + commission;
    }
}

/// <summary>
/// 63. Allocation of the parent class (Extract Superclass)
///
/// BEFORE: Duplicate code in classes
/// </summary>
public class DepartmentSuperBefore
{
    private string name;
    private string head;

    public DepartmentSuperBefore(string name, string head)
    {
        this.name = name;
        this.head = head;
    }

    public string GetName()
    {
        return name;
    }

    public string GetHead()
    {
        return head;
    }
}

public class CompanySuperBefore
{
    private string name;
    private string head;

    public CompanySuperBefore(string name, string head)
    {
        this.name = name;
        this.head = head;
    }

    public string GetName()
    {
        return name;
    }

    public string GetHead()
    {
        return head;
    }
}

/// <summary>
/// AFTER: Extract superclass
/// </summary>
public abstract class Party
{
    private string name;
    private string head;

    public Party(string name, string head)
    {
        this.name = name;
        this.head = head;
    }

    public string GetName()
    {
        return name;
    }

    public string GetHead()
    {
        return head;
    }
}

public class DepartmentSuperAfter : Party
{
    public DepartmentSuperAfter(string name, string head) : base(name, head) { }
}

public class CompanySuperAfter : Party
{
    public CompanySuperAfter(string name, string head) : base(name, head) { }
}

/// <summary>
/// 64. Interface extraction (Extract Interface)
///
/// BEFORE: Clients depend on concrete class
/// </summary>
public class PrinterInterfaceBefore
{
    public void Print(object document)
    {
        // Print logic
    }

    public string GetStatus()
    {
        // Status logic
        return "Ready";
    }

    public void CancelJob(int jobId)
    {
        // Cancel logic
    }
}

/// <summary>
/// AFTER: Extract interface
/// </summary>
public interface IPrinter
{
    void Print(object document);
    string GetStatus();
}

public class LaserPrinter : IPrinter
{
    public void Print(object document)
    {
        // Print logic
    }

    public string GetStatus()
    {
        // Status logic
        return "Ready";
    }

    public void CancelJob(int jobId)
    {
        // Cancel logic - not part of interface
    }
}

public class InkjetPrinter : IPrinter
{
    public void Print(object document)
    {
        // Print logic
    }

    public string GetStatus()
    {
        // Status logic
        return "Ready";
    }
}

/// <summary>
/// 65. Collapse Hierarchy
///
/// BEFORE: Unnecessary class hierarchy
/// </summary>
public class EmployeeCollapseBefore
{
}

public class ManagerCollapseBefore : EmployeeCollapseBefore
{
    private string department;
}

/// <summary>
/// AFTER: Collapse hierarchy if only one subclass
/// </summary>
public class EmployeeCollapseAfter
{
    private string department; // Moved up
}

/// <summary>
/// 66. Formation of the method template (Form Template Method)
///
/// BEFORE: Duplicate algorithm structure
/// </summary>
public class ReportGeneratorTemplateBefore
{
    public string GenerateHTMLReport()
    {
        var data = GetData();
        var header = FormatHeader();
        var body = FormatBody(data);
        var footer = FormatFooter();
        return header + body + footer;
    }

    public string GeneratePDFReport()
    {
        var data = GetData(); // Duplicate
        var header = FormatPDFHeader(); // Different
        var body = FormatPDFBody(data); // Different
        var footer = FormatPDFFooter(); // Different
        return header + body + footer;
    }

    protected virtual List<string> GetData()
    {
        return new List<string> { "item1", "item2" };
    }

    protected virtual string FormatHeader()
    {
        return "<h1>Report</h1>";
    }

    protected virtual string FormatBody(List<string> data)
    {
        return "<body>" + string.Join("", data) + "</body>";
    }

    protected virtual string FormatFooter()
    {
        return "<footer>End</footer>";
    }

    protected virtual string FormatPDFHeader()
    {
        return "PDF Report Header";
    }

    protected virtual string FormatPDFBody(List<string> data)
    {
        return "PDF Body: " + string.Join("", data);
    }

    protected virtual string FormatPDFFooter()
    {
        return "PDF Footer";
    }
}

/// <summary>
/// AFTER: Form template method
/// </summary>
public abstract class ReportGeneratorTemplateAfter
{
    public string GenerateReport()
    {
        var data = GetData();
        var header = FormatHeader();
        var body = FormatBody(data);
        var footer = FormatFooter();
        return AssembleReport(header, body, footer);
    }

    protected virtual List<string> GetData()
    {
        return new List<string> { "item1", "item2" };
    }

    protected abstract string FormatHeader();
    protected abstract string FormatBody(List<string> data);
    protected abstract string FormatFooter();
    protected abstract string AssembleReport(string header, string body, string footer);
}

public class HTMLReportGenerator : ReportGeneratorTemplateAfter
{
    protected override string FormatHeader()
    {
        return "<h1>Report</h1>";
    }

    protected override string FormatBody(List<string> data)
    {
        return "<body>" + string.Join("", data) + "</body>";
    }

    protected override string FormatFooter()
    {
        return "<footer>End</footer>";
    }

    protected override string AssembleReport(string header, string body, string footer)
    {
        return header + body + footer;
    }
}

public class PDFReportGenerator : ReportGeneratorTemplateAfter
{
    protected override string FormatHeader()
    {
        return "PDF Report Header";
    }

    protected override string FormatBody(List<string> data)
    {
        return "PDF Body: " + string.Join("", data);
    }

    protected override string FormatFooter()
    {
        return "PDF Footer";
    }

    protected override string AssembleReport(string header, string body, string footer)
    {
        return header + body + footer;
    }
}

/// <summary>
/// 67. Replacement of inheritance by delegation (Replace Inheritance with Delegation)
///
/// BEFORE: Inheritance where delegation would be better
/// </summary>
public class StackInheritanceBefore : List<object>
{
    public void Push(object item)
    {
        Add(item);
    }

    public object Pop()
    {
        if (Count == 0)
        {
            throw new InvalidOperationException("Stack is empty");
        }
        var lastIndex = Count - 1;
        var item = this[lastIndex];
        RemoveAt(lastIndex);
        return item;
    }
}

/// <summary>
/// AFTER: Replace inheritance with delegation
/// </summary>
public class StackDelegationAfter
{
    private List<object> items;

    public StackDelegationAfter()
    {
        items = new List<object>();
    }

    public void Push(object item)
    {
        items.Add(item);
    }

    public object Pop()
    {
        if (items.Count == 0)
        {
            throw new InvalidOperationException("Stack is empty");
        }
        var lastIndex = items.Count - 1;
        var item = items[lastIndex];
        items.RemoveAt(lastIndex);
        return item;
    }

    public int Count => items.Count;
}

/// <summary>
/// 68. Replacement of delegation by inheritance (Replace Delegation with Inheritance)
///
/// BEFORE: Delegation where inheritance would be simpler
/// </summary>
public class MyStringDelegateBefore
{
    private string str;

    public MyStringDelegateBefore(string str)
    {
        this.str = str;
    }

    public int Length()
    {
        return str.Length;
    }

    public string Substr(int start, int? length = null)
    {
        if (length.HasValue)
        {
            return str.Substring(start, length.Value);
        }
        return str.Substring(start);
    }

    public int IndexOf(string needle)
    {
        return str.IndexOf(needle);
    }
}

/// <summary>
/// AFTER: Replace delegation with inheritance
/// </summary>
public class MyStringInheritAfter : List<char>
{
    public MyStringInheritAfter(string str) : base(str.ToCharArray())
    {
    }

    public string ToString()
    {
        return new string(this.ToArray());
    }
}

File: nodejs/refactoring-methods/07-generalization-problems.js
/**
 * 57. Lifting the field (Pull Up Field)
 *
 * BEFORE: Duplicate fields in subclasses
 */
class EmployeePullBefore {
    constructor() {
        this.name = null;
    }
}

class ManagerPullBefore extends EmployeePullBefore {
    constructor() {
        super();
        this.name = null; // Duplicate
        this.budget = null;
    }
}

class EngineerPullBefore extends EmployeePullBefore {
    constructor() {
        super();
        this.name = null; // Duplicate
        this.skills = null;
    }
}

/**
 * AFTER: Pull up field to superclass
 */
class EmployeePullAfter {
    constructor() {
        this.name = null;
    }
}

class ManagerPullAfter extends EmployeePullAfter {
    constructor() {
        super();
        this.budget = null;
    }
}

class EngineerPullAfter extends EmployeePullAfter {
    constructor() {
        super();
        this.skills = null;
    }
}

/**
 * 58. Lifting the method (Pull Up Method)
 *
 * BEFORE: Duplicate methods in subclasses
 */
class ShapeMethodBefore {
    area() {
        throw new Error('Abstract method must be implemented');
    }
}

class CircleMethodBefore extends ShapeMethodBefore {
    constructor(radius) {
        super();
        this.radius = radius;
    }

    area() {
        return Math.PI * this.radius * this.radius;
    }

    circumference() {
        return 2 * Math.PI * this.radius;
    }
}

class SquareMethodBefore extends ShapeMethodBefore {
    constructor(side) {
        super();
        this.side = side;
    }

    area() {
        return this.side * this.side;
    }

    circumference() { // Duplicate logic
        return 4 * this.side;
    }
}

/**
 * AFTER: Pull up method to superclass
 */
class ShapeMethodAfter {
    area() {
        throw new Error('Abstract method must be implemented');
    }

    circumference() {
        throw new Error('Abstract method must be implemented');
    }
}

class CircleMethodAfter extends ShapeMethodAfter {
    constructor(radius) {
        super();
        this.radius = radius;
    }

    area() {
        return Math.PI * this.radius * this.radius;
    }

    circumference() {
        return 2 * Math.PI * this.radius;
    }
}

class SquareMethodAfter extends ShapeMethodAfter {
    constructor(side) {
        super();
        this.side = side;
    }

    area() {
        return this.side * this.side;
    }

    circumference() {
        return 4 * this.side;
    }
}

/**
 * 59. Lifting the constructor Body (Pull Up Constructor Body)
 *
 * BEFORE: Duplicate constructor code
 */
class VehicleConstructorBefore {
    constructor(make, model, year) {
        this.make = make;
        this.model = model;
        this.year = year;
    }
}

class CarConstructorBefore extends VehicleConstructorBefore {
    constructor(make, model, year, doors) {
        super(make, model, year); // Duplicate calls
        this.doors = doors;
    }
}

class TruckConstructorBefore extends VehicleConstructorBefore {
    constructor(make, model, year, payload) {
        super(make, model, year); // Duplicate calls
        this.payload = payload;
    }
}

/**
 * AFTER: Pull up constructor body (already properly structured in JS)
 */
class VehicleConstructorAfter {
    constructor(make, model, year) {
        this.make = make;
        this.model = model;
        this.year = year;
    }
}

class CarConstructorAfter extends VehicleConstructorAfter {
    constructor(make, model, year, doors) {
        super(make, model, year);
        this.doors = doors;
    }
}

class TruckConstructorAfter extends VehicleConstructorAfter {
    constructor(make, model, year, payload) {
        super(make, model, year);
        this.payload = payload;
    }
}

/**
 * 60. Method Descent (Push Down Method)
 *
 * BEFORE: Method in wrong class hierarchy level
 */
class AnimalPushBefore {
    speak() {
        // Generic implementation
        return 'Generic animal sound';
    }
}

class DogPushBefore extends AnimalPushBefore {
    speak() {
        return 'Woof';
    }
}

class CatPushBefore extends AnimalPushBefore {
    speak() {
        return 'Meow';
    }
}

class FishPushBefore extends AnimalPushBefore {
    // Fish don't speak, but inherits speak method
}

/**
 * AFTER: Push down method to appropriate subclasses
 */
class AnimalPushAfter {
    // No speak method here
}

class DogPushAfter extends AnimalPushAfter {
    speak() {
        return 'Woof';
    }
}

class CatPushAfter extends AnimalPushAfter {
    speak() {
        return 'Meow';
    }
}

class FishPushAfter extends AnimalPushAfter {
    // No speak method - appropriate for Fish
}

/**
 * 61. Field Descent (Push Down Field)
 *
 * BEFORE: Field in wrong hierarchy level
 */
class EmployeeFieldBefore {
    constructor() {
        this.salary = null; // Not all employees have salary
    }
}

class SalariedEmployeeFieldBefore extends EmployeeFieldBefore {
    // Uses salary
}

class ContractorFieldBefore extends EmployeeFieldBefore {
    // Doesn't use salary, but inherits it
}

/**
 * AFTER: Push down field
 */
class EmployeeFieldAfter {
    // No salary field
}

class SalariedEmployeeFieldAfter extends EmployeeFieldAfter {
    constructor() {
        super();
        this.salary = null;
    }
}

class ContractorFieldAfter extends EmployeeFieldAfter {
    constructor() {
        super();
        this.hourlyRate = null;
    }
}

/**
 * 62. Subclass extraction (Extract Subclass)
 *
 * BEFORE: Class with conditional behavior
 */
class JobExtractBefore {
    constructor(type, rate, commission = null) {
        this.type = type;
        this.rate = rate;
        this.commission = commission;
    }

    getPay() {
        if (this.type === 'salaried') {
            return this.rate;
        } else {
            return this.rate + this.commission;
        }
    }
}

/**
 * AFTER: Extract subclass
 */
class JobExtractAfter {
    constructor(rate) {
        this.rate = rate;
    }

    getPay() {
        throw new Error('Abstract method must be implemented');
    }
}

class SalariedJob extends JobExtractAfter {
    getPay() {
        return this.rate;
    }
}

class CommissionedJob extends JobExtractAfter {
    constructor(rate, commission) {
        super(rate);
        this.commission = commission;
    }

    getPay() {
        return this.rate + this.commission;
    }
}

/**
 * 63. Allocation of the parent class (Extract Superclass)
 *
 * BEFORE: Duplicate code in classes
 */
class DepartmentSuperBefore {
    constructor(name, head) {
        this.name = name;
        this.head = head;
    }

    getName() {
        return this.name;
    }

    getHead() {
        return this.head;
    }
}

class CompanySuperBefore {
    constructor(name, head) {
        this.name = name;
        this.head = head;
    }

    getName() {
        return this.name;
    }

    getHead() {
        return this.head;
    }
}

/**
 * AFTER: Extract superclass
 */
class Party {
    constructor(name, head) {
        this.name = name;
        this.head = head;
    }

    getName() {
        return this.name;
    }

    getHead() {
        return this.head;
    }
}

class DepartmentSuperAfter extends Party {}

class CompanySuperAfter extends Party {}

/**
 * 64. Interface extraction (Extract Interface)
 *
 * BEFORE: Clients depend on concrete class
 */
class PrinterInterfaceBefore {
    print(document) {
        // Print logic
    }

    getStatus() {
        // Status logic
    }

    cancelJob(jobId) {
        // Cancel logic
    }
}

/**
 * AFTER: Extract interface (using class as interface simulation)
 */
class Printer {
    print(document) {
        throw new Error('Abstract method must be implemented');
    }

    getStatus() {
        throw new Error('Abstract method must be implemented');
    }
}

class LaserPrinter extends Printer {
    print(document) {
        // Print logic
    }

    getStatus() {
        // Status logic
    }

    cancelJob(jobId) {
        // Cancel logic - not part of interface
    }
}

class InkjetPrinter extends Printer {
    print(document) {
        // Print logic
    }

    getStatus() {
        // Status logic
    }
}

/**
 * 65. Collapse Hierarchy
 *
 * BEFORE: Unnecessary class hierarchy
 */
class EmployeeCollapseBefore {
    constructor() {
        this.name = null;
    }
}

class ManagerCollapseBefore extends EmployeeCollapseBefore {
    constructor() {
        super();
        this.department = null;
    }
}

/**
 * AFTER: Collapse hierarchy if only one subclass
 */
class EmployeeCollapseAfter {
    constructor() {
        this.name = null;
        this.department = null; // Moved up
    }
}

/**
 * 66. Formation of the method template (Form Template Method)
 *
 * BEFORE: Duplicate algorithm structure
 */
class ReportGeneratorTemplateBefore {
    generateHTMLReport() {
        const data = this.getData();
        const header = this.formatHeader();
        const body = this.formatBody(data);
        const footer = this.formatFooter();
        return header + body + footer;
    }

    generatePDFReport() {
        const data = this.getData(); // Duplicate
        const header = this.formatPDFHeader(); // Different
        const body = this.formatPDFBody(data); // Different
        const footer = this.formatPDFFooter(); // Different
        return header + body + footer;
    }

    getData() {
        return ['item1', 'item2'];
    }

    formatHeader() {
        return '<h1>Report</h1>';
    }

    formatBody(data) {
        return '<body>' + data.join('') + '</body>';
    }

    formatFooter() {
        return '<footer>End</footer>';
    }

    formatPDFHeader() {
        return 'PDF Report Header';
    }

    formatPDFBody(data) {
        return 'PDF Body: ' + data.join('');
    }

    formatPDFFooter() {
        return 'PDF Footer';
    }
}

/**
 * AFTER: Form template method
 */
class ReportGeneratorTemplateAfter {
    generateReport() {
        const data = this.getData();
        const header = this.formatHeader();
        const body = this.formatBody(data);
        const footer = this.formatFooter();
        return this.assembleReport(header, body, footer);
    }

    getData() {
        return ['item1', 'item2'];
    }

    formatHeader() {
        throw new Error('Abstract method must be implemented');
    }

    formatBody(data) {
        throw new Error('Abstract method must be implemented');
    }

    formatFooter() {
        throw new Error('Abstract method must be implemented');
    }

    assembleReport(header, body, footer) {
        throw new Error('Abstract method must be implemented');
    }
}

class HTMLReportGenerator extends ReportGeneratorTemplateAfter {
    formatHeader() {
        return '<h1>Report</h1>';
    }

    formatBody(data) {
        return '<body>' + data.join('') + '</body>';
    }

    formatFooter() {
        return '<footer>End</footer>';
    }

    assembleReport(header, body, footer) {
        return header + body + footer;
    }
}

class PDFReportGenerator extends ReportGeneratorTemplateAfter {
    formatHeader() {
        return 'PDF Report Header';
    }

    formatBody(data) {
        return 'PDF Body: ' + data.join('');
    }

    formatFooter() {
        return 'PDF Footer';
    }

    assembleReport(header, body, footer) {
        return header + body + footer;
    }
}

/**
 * 67. Replacement of inheritance by delegation (Replace Inheritance with Delegation)
 *
 * BEFORE: Inheritance where delegation would be better
 */
class StackInheritanceBefore extends Array {
    push(item) {
        super.push(item);
    }

    pop() {
        if (this.length === 0) {
            throw new Error('Stack is empty');
        }
        return super.pop();
    }
}

/**
 * AFTER: Replace inheritance with delegation
 */
class StackDelegationAfter {
    constructor() {
        this.items = [];
    }

    push(item) {
        this.items.push(item);
    }

    pop() {
        if (this.items.length === 0) {
            throw new Error('Stack is empty');
        }
        return this.items.pop();
    }

    get length() {
        return this.items.length;
    }
}

/**
 * 68. Replacement of delegation by inheritance (Replace Delegation with Inheritance)
 *
 * BEFORE: Delegation where inheritance would be simpler
 */
class MyStringDelegateBefore {
    constructor(string) {
        this.string = string;
    }

    length() {
        return this.string.length;
    }

    substr(start, length = undefined) {
        return this.string.substr(start, length);
    }

    indexOf(needle) {
        return this.string.indexOf(needle);
    }
}

/**
 * AFTER: Replace delegation with inheritance
 */
class MyStringInheritAfter extends String {
    constructor(string) {
        super(string);
    }

    toMyString() {
        return this.toString();
    }
}

module.exports = {
    EmployeePullBefore,
    ManagerPullBefore,
    EngineerPullBefore,
    EmployeePullAfter,
    ManagerPullAfter,
    EngineerPullAfter,
    ShapeMethodBefore,
    CircleMethodBefore,
    SquareMethodBefore,
    ShapeMethodAfter,
    CircleMethodAfter,
    SquareMethodAfter,
    VehicleConstructorBefore,
    CarConstructorBefore,
    TruckConstructorBefore,
    VehicleConstructorAfter,
    CarConstructorAfter,
    TruckConstructorAfter,
    AnimalPushBefore,
    DogPushBefore,
    CatPushBefore,
    FishPushBefore,
    AnimalPushAfter,
    DogPushAfter,
    CatPushAfter,
    FishPushAfter,
    EmployeeFieldBefore,
    SalariedEmployeeFieldBefore,
    ContractorFieldBefore,
    EmployeeFieldAfter,
    SalariedEmployeeFieldAfter,
    ContractorFieldAfter,
    JobExtractBefore,
    JobExtractAfter,
    SalariedJob,
    CommissionedJob,
    DepartmentSuperBefore,
    CompanySuperBefore,
    Party,
    DepartmentSuperAfter,
    CompanySuperAfter,
    PrinterInterfaceBefore,
    Printer,
    LaserPrinter,
    InkjetPrinter,
    EmployeeCollapseBefore,
    ManagerCollapseBefore,
    EmployeeCollapseAfter,
    ReportGeneratorTemplateBefore,
    ReportGeneratorTemplateAfter,
    HTMLReportGenerator,
    PDFReportGenerator,
    StackInheritanceBefore,
    StackDelegationAfter,
    MyStringDelegateBefore,
    MyStringInheritAfter
};
